src/App/Fossa/Config/Common.hs

```diff
@@ -302,10 +302,16 @@ collectRevisionData (BaseDir basedir) maybeConfig cacheStrategy cliOverride = do
   let override = collectRevisionOverride maybeConfig cliOverride
   case cacheStrategy of
     ReadOnly -> do
-      inferred <- inferProjectFromVCS basedir <||> inferProjectCached basedir <||> inferProjectDefault basedir
+      maybeInferred <- inferProjectFromVCS basedir
+      inferred <- case maybeInferred of
+        Nothing -> inferProjectCached basedir <||> inferProjectDefault basedir
+        Just inferredFromVCS -> pure inferredFromVCS
       pure $ mergeOverride override inferred
     WriteOnly -> do
-      inferred <- inferProjectFromVCS basedir <||> inferProjectDefault basedir
+      maybeInferred <- inferProjectFromVCS basedir
+      inferred <- case maybeInferred of
```

Comment:
Case analysis can often be simplified with the [`maybe`](https://hackage.haskell.org/package/base-4.19.0.0/docs/Prelude.html#v:maybe) function:
```haskell
maybe :: b -> (a -> b) -> Maybe a -> b
```
The type describes what it does: It takes a default value, and a function that can be applied inside a `Maybe` to produce a value of that type. If the `Maybe a` is `Nothing` then it returns the default, otherwise it applies the function to `a`.

```haskell
inferred <- maybe (inferProjectDefault baseDir) pure (inferProjectFromVCS basedir)
```
should be equivalent. Keep in mind that Haskell is lazy, so `maybe` is able to return a value in the `Just` case without evaluating the default argument. If you look at the cases in its [source code](https://hackage.haskell.org/package/base-4.19.0.0/docs/src/Data.Maybe.html#maybe) it may be clearer why. 

Not directly useful here, but good to know is also the [`fromMaybe`](https://hackage.haskell.org/package/base-4.19.0.0/docs/Data-Maybe.html#v:fromMaybe) function.